presto-hive/src/main/java/com/facebook/presto/hive/S3SelectRecordCursor.java

```diff
@@ -0,0 +1,227 @@
+        checkArgument(parts[parts.length - 1].endsWith(END_STRUCT), "Invalid thrift DDL " + ddl);
+        parts[parts.length - 1] = parts[parts.length - 1].substring(0, parts[parts.length - 1].length() - 1);
+        List<ThriftField> fields = new ArrayList<>();
+        for (int i = 3; i < parts.length - 1; i += 2) {
+            ThriftField thriftField = new ThriftField();
+            thriftField.setType(parts[i]);
+            if (parts[i + 1].endsWith(FIELD_SEPARATOR)) {
+                parts[i + 1] = parts[i + 1].substring(0, parts[i + 1].length() - 1);
+            }
+            thriftField.setName(parts[i + 1]);
+            fields.add(thriftField);
+        }
+        thriftDDL.setFields(fields);
+
+        return thriftDDL;
+    }
+
+    private static ThriftDDL filterThriftDDL(ThriftDDL thriftDDL, List<HiveColumnHandle> columns)
+    {
+        if (thriftDDL == null) {
+            return null;
+        }
+        List<ThriftField> fields = thriftDDL.getFields();
+        if (fields == null || fields.isEmpty()) {
+            return thriftDDL;
+        }
+        Set<String> columnNames = columns.stream()
+                .map(HiveColumnHandle::getName)
+                .collect(toImmutableSet());
+        List<ThriftField> filteredFields = fields.stream()
+                .filter(field -> columnNames.contains(field.getName()))
+                .collect(toList());
+        thriftDDL.setFields(filteredFields);
+
+        return thriftDDL;
+    }
+
+    private static String thriftDDLToString(ThriftDDL thriftDDL)
+    {
+        if (thriftDDL == null) {
+            return EMPTY_STRING;
+        }
+        List<ThriftField> fields = thriftDDL.getFields();
+        if (fields == null || fields.isEmpty()) {
+            return EMPTY_STRING;
+        }
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(THRIFT_STRUCT)
+                .append(SPACE)
+                .append(thriftDDL.getTableName())
+                .append(SPACE)
+                .append(START_STRUCT);
+        stringBuilder.append(fields.stream()
+                .map(field -> SPACE + field.getType() + SPACE + field.getName())
+                .collect(Collectors.joining(",")));
+        stringBuilder.append(END_STRUCT);
+
+        return stringBuilder.toString();
+    }
+
+    private static class ThriftField
+    {
+        private String type;
+        private String name;
+
+        private String getType()
+        {
+            return type;
+        }
+
+        private void setType(String type)
+        {
+            checkArgument(!isNullOrEmpty(type), "type is null or empty string");
+            this.type = type;
+        }
+
+        private String getName()
+        {
+            return name;
+        }
+
+        private void setName(String name)
+        {
+            requireNonNull(name, "name is null");
+            this.name = name;
+        }
+    }
+
+    private static class ThriftDDL
```

Comment:
`ThriftTable`?

Reply:
It's not a DDL and is a representation of table. Refactoring this to ThriftTable would be better.